test/checkChallenge.test.js

```diff
@@ -286,6 +287,111 @@ describe('Challenges', () => {
    
   });
 
+  describe('Rejectnew', () => {
+
+    it('reject all games if rejectnew is used with default rejectnewmsg', () => {
+
+      const notification = base_challenge({ ranked: false });
+
+      config.rejectnew = true;
+
+      const result = conn.checkChallengeBot(notification, fs);
```

Comment:
fs is now an external parameter that is passed to checkChallengeBot, so in the future if we want to write a fake fs we can easily do it in checkChallenge.test.js

should i add mock-fs (https://www.npmjs.com/package/mock-fs) to mock fs too, or is it fine to keep default fs in checkChallenge.test.js for now?

the only other occurence of fs in all our code is in console.js, so when we write testing for console.js, it may be a good time to make fs an external parameter too

Reply:
also i tried to write manually a FakeFs class similar to FakeSocket but i was unable to do it :sweat_smile:

Reply:
oh actually it seems mock-fs is not used to simulate a filesystem's paths architecture, but it does not provide fs methods such as fs.existsSync

i read on the fs doc (https://nodejs.org/api/fs.html#fs_file_descriptors) that fs abstracts filepath to be operating system aspecific, so if we find a way to write a FakeFs that has our own fake method existsSync, would it work on all operating systems (ex: windows)?

Reply:
You can always use a sinon spy. However since right now we're only reading the file there is no direct need to do so yet. In fact using the real filesystem is acceptable for testing. These tests are usually called integration tests or end to end tests.

When you do have side-effects, you will need to make sure the test's initalise and clean up, do not collide etc. etc. Also we need to consider performance for the tests. For performance reasons and also test stability (test become very fragile when they all try to access the same resource and all need to initialise/clean up), it's usual to write unit-tests and those unit-test will instead use a fake/mocked instance of the io provider. This has the advantage you use a local instance and have no side-effects.

For the current tests we only read the files so it's actually ok to keep it as is. When we have a test that writes to the log, we must switch to a fake instead.

Reply:
ah i see, i get the main idea

so if we want to write testing for console.js, then we'd need to write a fake fs too at that time

so until then, is this pr good to merge, or should we take the opportunity to also write a fake fs (i'm interested how that can be done if we can do it in this pr)

Reply:
One other thing: Because of the way node imports work, it might be possible to replace the result of `require` calls in our tests, removing the need to insert includes like this in the constructor.

Reply:
ah but i didnt insert fs in the constructor, i tried to insert it in the constructor at first but it was uneeded to pass fs for all of our code if we only use it once in check_rejectnew

so i refactored this part of the code so that each checkChallenge subfonction can have its own parameters, not just notification:
https://github.com/online-go/gtp2ogs/pull/359/files#diff-0d74c7455b8cbe5be6f644690ecb1b8aR530
https://github.com/online-go/gtp2ogs/pull/359/files#diff-0d74c7455b8cbe5be6f644690ecb1b8aR385
https://github.com/online-go/gtp2ogs/pull/359/files#diff-0d74c7455b8cbe5be6f644690ecb1b8aR805


thats why in our tests we can pass fs only to checkChallengeBot, but we dont need to pass it in the connection constructor, nor in any other checkChallenge subfonction:
https://github.com/online-go/gtp2ogs/pull/359/files#diff-877056d4e4a6a8bc8dca4776a268c187R298

so gtp2ogs uses native fs, but if we want in our test we can pass a fake fs too now

i find it a lot nicer this way, what do you think

Reply:
i would still like to see how to fake fs (with sinon or some other way) if we can before merging this pr :)

Reply:
but since challenge always checks notification, we could slightly optimize it by writing notification only once and then the rest of params is everything other than notification